File: lib/services/auth_service.dart
import 'package:flutter/material.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/user_type.dart';
import 'dart:developer' as developer;

class AuthService extends ChangeNotifier {
  final SupabaseClient _supabase = Supabase.instance.client;
  String? _error;
  bool _isLoading = false;
  UserType? _userType;
  User? _currentUser;
  String? _userName;

  bool get isLoading => _isLoading;
  String? get error => _error;
  UserType? get userType => _userType;
  User? get currentUser => _currentUser;
  String? get userName => _userName;

  Future<void> initialize() async {
    Supabase.instance.client.auth.onAuthStateChange.listen((data) {
      final AuthChangeEvent event = data.event;
      final Session? session = data.session;

      if (event == AuthChangeEvent.signedIn && session != null) {
        _createInitialProfile(session.user);
      }
    });
  }

  Future<void> _createInitialProfile(User user) async {
    try {
      print('Verificando/criando perfil para usuário ${user.id}');

      final existingProfile = await _supabase
          .from('profiles')
          .select()
          .eq('id', user.id)
          .maybeSingle();

      if (existingProfile == null) {
        print('Perfil não encontrado, criando novo perfil...');

        await _supabase.from('profiles').insert({
          'id': user.id,
          'email': user.email,
          'name': user.email?.split('@')[0] ?? 'Usuário',
          'user_type': 'aluno'
        });

        print('Perfil criado com sucesso');
      } else {
        print('Perfil já existe');
      }
    } catch (e, stackTrace) {
      print('=== ERRO AO CRIAR PERFIL ===');
      print('Erro: $e');
      print('Stack trace: $stackTrace');
    }
  }

  Future<bool> login(String email, String password) async {
    try {
      _error = null;
      _isLoading = true;
      notifyListeners();

      print('=== INICIANDO LOGIN ===');
      print('Email: $email');

      final response = await _supabase.auth.signInWithPassword(
        email: email,
        password: password,
      );

      if (response.user == null) {
        print('Login falhou: usuário é null');
        _error = 'Usuário ou senha inválidos';
        return false;
      }

      print('Login bem sucedido. ID do usuário: ${response.user!.id}');
      _currentUser = response.user;

      print('Verificando perfil existente...');
      final existingProfile = await _supabase
          .from('profiles')
          .select()
          .eq('id', response.user!.id)
          .maybeSingle();

      if (existingProfile == null) {
        print('Perfil não encontrado, criando novo perfil...');
        final newProfile = {
          'id': response.user!.id,
          'email': response.user!.email,
          'name': response.user!.email?.split('@')[0] ?? 'Usuário',
          'user_type': 'aluno'
        };

        await _supabase.from('profiles').insert(newProfile);
        print('Novo perfil criado com sucesso');

        _userType = UserType.aluno;
        _userName = newProfile['name'];
      } else {
        print('Perfil encontrado: $existingProfile');
        _userType = UserType.fromString(existingProfile['user_type']);
        _userName = existingProfile['name'];
      }

      print('=== LOGIN COMPLETO ===');
      print('Nome: $_userName');
      print('Tipo: ${_userType?.name}');

      return true;
    } catch (e, stackTrace) {
      print('=== ERRO NO LOGIN ===');
      print('Erro: $e');
      print('Stack trace: $stackTrace');
      _error = _handleError(e);
      return false;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<bool> register(
      String name, String email, String password, UserType userType) async {
    try {
      _error = null;
      _isLoading = true;
      notifyListeners();

      print('=== INICIANDO REGISTRO ===');
      print('Email: $email');

      final existingUser = await _supabase
          .from('profiles')
          .select()
          .eq('email', email)
          .maybeSingle();

      if (existingUser != null) {
        print('Email já registrado');
        _error = 'Este email já está cadastrado no sistema';
        return false;
      }

      print('Email disponível, prosseguindo com registro...');

      final response = await _supabase.auth.signUp(
        email: email,
        password: password,
        data: {'name': name, 'user_type': userType.name},
      );

      if (response.user == null) {
        print('Falha ao criar usuário no Auth');
        _error = 'Falha ao criar usuário';
        return false;
      }

      print('Usuário criado no Auth com sucesso. ID: ${response.user!.id}');

      try {
        await _supabase.from('profiles').upsert({
          'id': response.user!.id,
          'name': name,
          'email': email,
          'user_type': userType.name,
          'created_at': DateTime.now().toIso8601String(),
          'updated_at': DateTime.now().toIso8601String(),
        }, onConflict: 'id');

        print('Perfil criado/atualizado com sucesso na tabela profiles');

        _currentUser = response.user;
        _userType = userType;
        _userName = name;

        print('=== REGISTRO COMPLETO COM SUCESSO ===');
        return true;
      } catch (profileError) {
        print('Erro ao criar perfil: $profileError');
        await _supabase.auth.signOut();
        throw profileError;
      }
    } catch (e, stackTrace) {
      print('=== ERRO NO REGISTRO ===');
      print('Erro: $e');
      print('Stack trace: $stackTrace');
      _error = _handleError(e);

      try {
        await _supabase.auth.signOut();
      } catch (e) {
        print('Erro ao fazer logout: $e');
      }

      return false;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<bool> resetPassword(String email) async {
    try {
      _error = null;
      _isLoading = true;
      notifyListeners();

      await _supabase.auth.resetPasswordForEmail(
        email,
        redirectTo: 'io.supabase.orbirq://reset-callback/',
      );

      return true;
    } catch (e) {
      _error = _handleError(e);
      return false;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<void> logout() async {
    try {
      await _supabase.auth.signOut();
      _currentUser = null;
      _userType = null;
    } catch (e) {
      _error = _handleError(e);
    }
    notifyListeners();
  }

  Future<bool> checkCurrentSession() async {
    try {
      final session = _supabase.auth.currentSession;

      if (session != null) {
        _currentUser = session.user;

        final userData = await _supabase
            .from('profiles')
            .select('user_type, name')
            .eq('id', session.user!.id)
            .limit(1) 
            .maybeSingle();

        if (userData != null) {
          _userType = UserType.fromString(userData['user_type']);
          _userName = userData['name'];
          await _createInitialProfile(session.user!);
          return true;
        } else {
          _error = 'Perfil do usuário não encontrado';
          return false;
        }
      }
      return false;
    } catch (e) {
      _error = _handleError(e);
      return false;
    } finally {
      notifyListeners();
    }
  }

  String _handleError(dynamic e) {
    if (e is AuthException) {
      switch (e.message) {
        case 'Invalid login credentials':
          return 'Email ou senha inválidos';
        case 'Email not confirmed':
          return 'Por favor, confirme seu email antes de fazer login';
        case 'User already registered':
          return 'Este email já está cadastrado no sistema';
        case 'Password should be at least 6 characters':
          return 'A senha deve ter pelo menos 6 caracteres';
        case 'Invalid email':
          return 'O email fornecido é inválido';
        case 'Email rate limit exceeded':
          return 'Muitas tentativas. Por favor, aguarde alguns minutos antes de tentar novamente';
        case 'Signup requires a valid password':
          return 'É necessário fornecer uma senha válida';
        case 'Unable to validate email address: invalid format':
          return 'O formato do email é inválido';
        default:
          print('Erro original: ${e.message}'); 
          return 'Erro no cadastro: ${e.message}';
      }
    } else if (e is PostgrestException) {
      switch (e.code) {
        case '23505': 
          return 'Este usuário já existe no sistema';
        case '23502': 
          return 'Por favor, preencha todos os campos obrigatórios';
        default:
          print('Erro Postgres: ${e.message}'); 
          return 'Erro no banco de dados: ${e.message}';
      }
    }
    print('Erro não categorizado: $e');
    return 'Ocorreu um erro inesperado. Por favor, tente novamente.';
  }
}

File: website/src/services/auth.service.ts
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL!
const supabaseKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!

export const supabase = createClient(supabaseUrl, supabaseKey)

export type UserType = 'aluno' | 'professor'

export interface Profile {
  id: string
  email: string
  name: string
  user_type: UserType
  created_at?: string
  updated_at?: string
}

class AuthService {
  private static instance: AuthService
  private _currentUser: Profile | null = null
  private _isLoading = false
  private _error: string | null = null

  private constructor() {}

  static getInstance(): AuthService {
    if (!AuthService.instance) {
      AuthService.instance = new AuthService()
    }
    return AuthService.instance
  }

  get currentUser() {
    return this._currentUser
  }

  get isLoading() {
    return this._isLoading
  }

  get error() {
    return this._error
  }

  get supabase() {
    return supabase
  }

  async initialize() {
    // Adiciona listener para mudanças no estado de autenticação
    supabase.auth.onAuthStateChange(async (event, session) => {
      if (event === 'SIGNED_IN' && session) {
        await this._createInitialProfile(session.user)
      }
    })

    // Verifica sessão atual
    await this.checkCurrentSession()
  }

  private async _createInitialProfile(user: any) {
    try {
      console.log('Verificando/criando perfil para usuário', user.id)

      // Verifica se o perfil já existe
      const { data: existingProfile } = await supabase
        .from('profiles')
        .select()
        .eq('id', user.id)
        .maybeSingle()

      if (!existingProfile) {
        console.log('Perfil não encontrado, criando novo perfil...')

        // Cria o perfil se não existir
        await supabase.from('profiles').insert({
          id: user.id,
          email: user.email,
          name: user.email?.split('@')[0] ?? 'Usuário',
          user_type: 'aluno'
        })

        console.log('Perfil criado com sucesso')
      } else {
        console.log('Perfil já existe')
      }
    } catch (e) {
      console.error('Erro ao criar perfil:', e)
    }
  }

  async login(email: string, password: string): Promise<boolean> {
    try {
      this._error = null
      this._isLoading = true

      console.log('=== INICIANDO LOGIN ===')
      console.log('Email:', email)

      // 1. Tentar fazer login
      const { data: authData, error: authError } = await supabase.auth.signInWithPassword({
        email,
        password,
      })

      if (authError || !authData.user) {
        console.log('Login falhou:', authError?.message)
        this._error = 'Usuário ou senha inválidos'
        return false
      }

      console.log('Login bem sucedido. ID do usuário:', authData.user.id)

      // 2. Verificar se o perfil existe
      console.log('Verificando perfil existente...')
      const { data: profile } = await supabase
        .from('profiles')
        .select()
        .eq('id', authData.user.id)
        .maybeSingle()

      if (!profile) {
        console.log('Perfil não encontrado, criando novo perfil...')
        // 3. Criar perfil se não existir
        const newProfile: Profile = {
          id: authData.user.id,
          email: authData.user.email!,
          name: authData.user.email?.split('@')[0] ?? 'Usuário',
          user_type: 'aluno'
        }

        await supabase.from('profiles').insert(newProfile)
        console.log('Novo perfil criado com sucesso')

        this._currentUser = newProfile
      } else {
        console.log('Perfil encontrado:', profile)
        this._currentUser = profile
      }

      return true
    } catch (e) {
      console.error('=== ERRO NO LOGIN ===', e)
      this._error = this._handleError(e)
      return false
    } finally {
      this._isLoading = false
    }
  }

  async register(name: string, email: string, password: string, userType: UserType): Promise<boolean> {
    try {
      this._error = null
      this._isLoading = true

      console.log('=== INICIANDO REGISTRO ===')
      console.log('Email:', email)

      // 1. Verificar se já existe uma conta
      const { data: existingUser } = await supabase
        .from('profiles')
        .select()
        .eq('email', email)
        .maybeSingle()

      if (existingUser) {
        console.log('Email já registrado')
        this._error = 'Este email já está cadastrado no sistema'
        return false
      }

      console.log('Email disponível, prosseguindo com registro...')

      // 2. Realizar o cadastro do usuário no Auth
      const { data: authData, error: authError } = await supabase.auth.signUp({
        email,
        password,
        options: {
          data: { name, user_type: userType }
        }
      })

      if (authError || !authData.user) {
        console.log('Falha ao criar usuário no Auth')
        this._error = 'Falha ao criar usuário'
        return false
      }

      console.log('Usuário criado no Auth com sucesso. ID:', authData.user.id)

      // 3. Criar ou atualizar o perfil na tabela profiles
      const profile: Profile = {
        id: authData.user.id,
        name,
        email,
        user_type: userType,
      }

      const { error: profileError } = await supabase
        .from('profiles')
        .upsert(profile)

      if (profileError) {
        console.log('Erro ao criar perfil:', profileError)
        await supabase.auth.signOut()
        throw profileError
      }

      console.log('Perfil criado/atualizado com sucesso')
      this._currentUser = profile

      return true
    } catch (e) {
      console.error('=== ERRO NO REGISTRO ===', e)
      this._error = this._handleError(e)
      await supabase.auth.signOut()
      return false
    } finally {
      this._isLoading = false
    }
  }

  async resetPassword(email: string): Promise<boolean> {
    try {
      this._error = null
      this._isLoading = true

      const { error } = await supabase.auth.resetPasswordForEmail(email, {
        redirectTo: `${window.location.origin}/reset-password`,
      })

      if (error) throw error
      return true
    } catch (e) {
      this._error = this._handleError(e)
      return false
    } finally {
      this._isLoading = false
    }
  }

  async logout(): Promise<void> {
    try {
      await supabase.auth.signOut()
      this._currentUser = null
    } catch (e) {
      this._error = this._handleError(e)
    }
  }

  async checkCurrentSession(): Promise<boolean> {
    try {
      const { data: { session } } = await supabase.auth.getSession()

      if (session?.user) {
        // Buscar o perfil completo do usuário
        const { data: profile } = await supabase
          .from('profiles')
          .select()
          .eq('id', session.user.id)
          .maybeSingle()

        if (profile) {
          this._currentUser = profile
          await this._createInitialProfile(session.user)
          return true
        } else {
          this._error = 'Perfil do usuário não encontrado'
          return false
        }
      }
      return false
    } catch (e) {
      this._error = this._handleError(e)
      return false
    }
  }

  private _handleError(e: any): string {
    console.error('Erro original:', e)

    if (e?.message) {
      switch (e.message) {
        case 'Invalid login credentials':
          return 'Email ou senha inválidos'
        case 'Email not confirmed':
          return 'Por favor, confirme seu email antes de fazer login'
        case 'User already registered':
          return 'Este email já está cadastrado no sistema'
        case 'Password should be at least 6 characters':
          return 'A senha deve ter pelo menos 6 caracteres'
        case 'Invalid email':
          return 'O email fornecido é inválido'
        default:
          return `Erro: ${e.message}`
      }
    }

    return 'Ocorreu um erro inesperado. Por favor, tente novamente.'
  }
}

export const authService = AuthService.getInstance() 